# 1 Introduction and clinical context

# 2 Defintions methods and hypotheses

# 3 Import of data and libraries

## 3.1 Import of libraries

In [6]:
#Importing basic data analyst libraries
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

#Importing libraries for SQL querrying
import duckdb 

## 3.2 Import of tables

I decided to import all the tables relevant for my analysis at this point. I will treat them as a database and later on i will look at them in detail. Now I will just import them one by one whereas querrying and cleaning comes a bit later in separate stages. I chose this approach because i find it more logically organised compared to other solutions.

###3.2.1 Importing confirmed cases table

In [10]:
df_cases = pd.read_csv('data/raw/confirmed_global.csv')

Verifying import

In [12]:
print(df_cases.head())

  Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  4/1/21  4/2/21  4/3/21  4/4/21  \
0        0        0        0        0  ...   56517   56572   56595   56676   
1        0        0        0        0  ...  125506  125842  126183  126531   
2        0        0        0        0  ...  117304  117429  117524  117622   
3        0        0        0        0  ...   12053   12115   12174   12231   
4        0        0        0        0  ...   22399   22467   22579   22631   

   4/5/21  4/6/21  4/7/21  4/8/21  4/9/21  4/10/21  
0   56717   56779   56873   56943

##### ---Cases dataset imported successfully---

### 3.2.2 Importing death cases

In [15]:
df_deaths = pd.read_csv('data/raw/deaths_global.csv')

Verifying import

In [17]:
print(df_deaths.head())

  Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  4/1/21  4/2/21  4/3/21  4/4/21  \
0        0        0        0        0  ...    2489    2495    2496    2497   
1        0        0        0        0  ...    2241    2247    2256    2265   
2        0        0        0        0  ...    3096    3099    3102    3105   
3        0        0        0        0  ...     115     116     117     117   
4        0        0        0        0  ...     538     538     540     542   

   4/5/21  4/6/21  4/7/21  4/8/21  4/9/21  4/10/21  
0    2508    2512    2512    2516

##### ----Deaths dataset importet successfully----

### 3.2.3 Importing recovered cases dataset

In [20]:
df_recovered = pd.read_csv('data/raw/recovered_global.csv')

In [21]:
print(df_recovered.head())

  Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  4/1/21  4/2/21  4/3/21  4/4/21  \
0        0        0        0        0  ...   51788   51798   51802   51885   
1        0        0        0        0  ...   91875   92500   93173   93842   
2        0        0        0        0  ...   81632   81729   81813   81896   
3        0        0        0        0  ...   11365   11401   11428   11474   
4        0        0        0        0  ...   20508   20867   20871   20879   

   4/5/21  4/6/21  4/7/21  4/8/21  4/9/21  4/10/21  
0   51902   51928   51940   51956

##### ----Recovered dataset importet successfully!----

# 4 Inspecting cleaning and preparing datasets for melting

In this sections I will get acquianted with dataset, melt it, deal with missings and prepare it for melting

## 4.1 Inspecting cleaning and preparing the dataset with cases

### 4.1.1 Inspecting the dataset with cases

In [27]:
print(df_cases.head())

  Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  ...  4/1/21  4/2/21  4/3/21  4/4/21  \
0        0        0        0        0  ...   56517   56572   56595   56676   
1        0        0        0        0  ...  125506  125842  126183  126531   
2        0        0        0        0  ...  117304  117429  117524  117622   
3        0        0        0        0  ...   12053   12115   12174   12231   
4        0        0        0        0  ...   22399   22467   22579   22631   

   4/5/21  4/6/21  4/7/21  4/8/21  4/9/21  4/10/21  
0   56717   56779   56873   56943

In [28]:
print(df_cases.shape)

(274, 449)


Shape of the dataset is 274 rows and 449 columns. 

From the first glance I see identifier columns such as province/state, Country/Relogion, Latitude and longitude. There are 449 columns where majority corresponds to individual dates within given period of time. To confirm that only date columns follow 'Long', and that there are no unexpected columns in between, I will iterate through the column names and count the columns to crosscheck whether i itterate through columns correctly.

In [31]:
count = 0
for col in df_cases:
    print(col)
    count = count + 1
    
print("the column count is: ", count)

Province/State
Country/Region
Lat
Long
1/22/20
1/23/20
1/24/20
1/25/20
1/26/20
1/27/20
1/28/20
1/29/20
1/30/20
1/31/20
2/1/20
2/2/20
2/3/20
2/4/20
2/5/20
2/6/20
2/7/20
2/8/20
2/9/20
2/10/20
2/11/20
2/12/20
2/13/20
2/14/20
2/15/20
2/16/20
2/17/20
2/18/20
2/19/20
2/20/20
2/21/20
2/22/20
2/23/20
2/24/20
2/25/20
2/26/20
2/27/20
2/28/20
2/29/20
3/1/20
3/2/20
3/3/20
3/4/20
3/5/20
3/6/20
3/7/20
3/8/20
3/9/20
3/10/20
3/11/20
3/12/20
3/13/20
3/14/20
3/15/20
3/16/20
3/17/20
3/18/20
3/19/20
3/20/20
3/21/20
3/22/20
3/23/20
3/24/20
3/25/20
3/26/20
3/27/20
3/28/20
3/29/20
3/30/20
3/31/20
4/1/20
4/2/20
4/3/20
4/4/20
4/5/20
4/6/20
4/7/20
4/8/20
4/9/20
4/10/20
4/11/20
4/12/20
4/13/20
4/14/20
4/15/20
4/16/20
4/17/20
4/18/20
4/19/20
4/20/20
4/21/20
4/22/20
4/23/20
4/24/20
4/25/20
4/26/20
4/27/20
4/28/20
4/29/20
4/30/20
5/1/20
5/2/20
5/3/20
5/4/20
5/5/20
5/6/20
5/7/20
5/8/20
5/9/20
5/10/20
5/11/20
5/12/20
5/13/20
5/14/20
5/15/20
5/16/20
5/17/20
5/18/20
5/19/20
5/20/20
5/21/20
5/22/20
5/23/20
5/24/20
5/25/

All the respective columns  after 'Long' seem to represent individual dates and column count from the loop is in agreement with the column count from df_cases.head(). Next I need to check data types of individual columns

In [33]:
count_integer = 0
count_other_datatypes = 0
for col in df_cases:
    print(df_cases[col].dtype)
    if df_cases[col].dtype == int:
        count_integer = count_integer +1
    else:
        count_other_datatypes = count_other_datatypes + 1

print('Integer count is:', count_integer, '\n Other data types count is:', count_other_datatypes)
if count_integer == 445:
    print('All date-columns are integers!')

object
object
float64
float64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int6

Loop confirmed that all date columns are formated as an integer datatype. I see that other datatypes in the dataset are objects and float. I will confirm it with command .info()

In [35]:
df_cases.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Columns: 449 entries, Province/State to 4/10/21
dtypes: float64(2), int64(445), object(2)
memory usage: 961.3+ KB


Verification showed exactly those datatypes i noticed  above

In [37]:
print(df_cases.describe)

<bound method NDFrame.describe of     Province/State      Country/Region        Lat        Long  1/22/20  \
0              NaN         Afghanistan  33.939110   67.709953        0   
1              NaN             Albania  41.153300   20.168300        0   
2              NaN             Algeria  28.033900    1.659600        0   
3              NaN             Andorra  42.506300    1.521800        0   
4              NaN              Angola -11.202700   17.873900        0   
..             ...                 ...        ...         ...      ...   
269            NaN             Vietnam  14.058324  108.277199        0   
270            NaN  West Bank and Gaza  31.952200   35.233200        0   
271            NaN               Yemen  15.552727   48.516388        0   
272            NaN              Zambia -13.133897   27.849332        0   
273            NaN            Zimbabwe -19.015438   29.154857        0   

     1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  ...  4/1/21  4/2/21  4/3/21

Describe command gives me an overview over the dataset. There are a lot of missing values in province/state. I have to investigate this variable closer. Country/region column will likely be the variable of interest and will serve us as identifier, Whereas latitute and longitude are not interesting forour analysis. Even though some viruses are known to spread better in colder or warmer climates, Cross-country comparisons in cases between the countries are not feasible because we don´t have data on people tested and countries were very different in test measures and  test equipment. In this dataset we lack the data for it. I will not check dataset with .describe(). Quick overview shows that there were 0 cases in a begining whereas number of cases began to increase and fluctuate over time.

### 4.1.2 Cleaning the dataset with cases

I am checking if there are any missing values in columns

In [41]:
df_cases.isna()

Province/State  Country/Region    Lat   Long  1/22/20  1/23/20  1/24/20  \
0              True           False  False  False    False    False    False   
1              True           False  False  False    False    False    False   
2              True           False  False  False    False    False    False   
3              True           False  False  False    False    False    False   
4              True           False  False  False    False    False    False   
..              ...             ...    ...    ...      ...      ...      ...   
269            True           False  False  False    False    False    False   
270            True           False  False  False    False    False    False   
271            True           False  False  False    False    False    False   
272            True           False  False  False    False    False    False   
273            True           False  False  False    False    False    False   

     1/25/20  1/26/20  1/27/20  ...  4/1/21  4/2/21  4/3/21  4/4/21  4/5/21  \
0      False    False    False  ...   False   False   False   False   False   
1      False    False    False  ...   False   False   False   False   False   
2      False    False    False  ...   False   False   False   False   False   
3      False    False    False  ...   False   False   False   False   False   
4      False    False    False  ...   False   False   False   False   False   
..       ...      ...      ...  ...     ...     ...     ...     ...     ...   
269    False    False    False  ...   False   False   False   False   False   
270    False    False    False  ...   False   False   False   False   False   
271    False    False    False  ...   False   False   False   False   False   
272    False    False    False  ...   False   False   False   False   False   
273    False    False    False  ...   False   False   False   False   False   

     4/6/21  4/7/21  4/8/21  4/9/21  4/10/21  
0     False   False   False   False    False  
1     False   False   False   False    False  
2     False   False   False   False    False  
3     False   False   False   False    False  
4     False   False   False   False    False  
..      ...     ...     ...     ...      ...  
269   False   False   False   False    False  
270   False   False   False   False    False  
271   False   False   False   False    False  
272   False   False   False   False    False  
273   False   False   False   False    False  

[274 rows x 449 columns]

I cannot see the whole dataset, but the output indicates that Province/state has missing values. I am going to investigate it further along with other columns

In [43]:
df_cases.isna().any()

Province/State     True
Country/Region    False
Lat                True
Long               True
1/22/20           False
                  ...  
4/6/21            False
4/7/21            False
4/8/21            False
4/9/21            False
4/10/21           False
Length: 449, dtype: bool

It seems that longitude latitude and province contain misssing values whereas dates seem to contain no missing value. I need to check my assumption.

In [45]:
df_cases.isna().any().sum()


3

There are three columns in the df_cases. They are Province/State, Lat, Long.

There are three columns in the df_cases. They are Province/State, Lat, Long.
Before I go further with cleaning and droping, want to check which column is the most pertinent one to become identifier column during melting. The two most valiable options are Province/state and Country/Region. I will extend the jupyter notebook output to contain up to 500 rows and columns which should be sufficient size to cover the entire df_cases dataframe.

In [48]:
#extending output for rows
pd.set_option('display.max_rows', 300)
#extending output for columns
pd.set_option('display.max_columns', 500)

In [49]:
print(df_cases[['Province/State', 'Country/Region']])

                                   Province/State  \
0                                             NaN   
1                                             NaN   
2                                             NaN   
3                                             NaN   
4                                             NaN   
5                                             NaN   
6                                             NaN   
7                                             NaN   
8                    Australian Capital Territory   
9                                 New South Wales   
10                             Northern Territory   
11                                     Queensland   
12                                South Australia   
13                                       Tasmania   
14                                       Victoria   
15                              Western Australia   
16                                            NaN   
17                                            

I confirm that column 'Province/State' is of no interest as it shows either overseas territories and provinces/states within a few countries.Contrary, Norway and Slovakia are both within Country/Region column which will become our identifier. Now I will strip the dataset from all the unneccesarry columns and rows.

In [51]:
df_cases_filtered = """
SELECT *
FROM df_cases
WHERE "Country/Region" = 'Norway' OR "Country/Region" = 'Slovakia';
"""
df_cases = duckdb.query(df_cases_filtered).to_df()

Checking whether i filtered rows correctly

In [53]:
print(df_cases.iloc[:, 1:])
print(type(df_cases), df_cases.shape)



  Country/Region     Lat     Long  1/22/20  1/23/20  1/24/20  1/25/20  \
0         Norway  60.472   8.4689        0        0        0        0   
1       Slovakia  48.669  19.6990        0        0        0        0   

   1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  \
0        0        0        0        0        0        0       0       0   
1        0        0        0        0        0        0       0       0   

   2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  \
0       0       0       0       0       0       0       0        0        0   
1       0       0       0       0       0       0       0        0        0   

   2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  \
0      

I correctly filtered just relevant rows and transformed querried data into dataframe. There are two rows- Slovakia and Norway with number of columns(449) remained unchanged. Now I will proceed with dropping all the unneccessarry columns. namely: 
'Province/State', 
'Long', 
'Lat'

As mentioned earlier, variables lattitude and longitude are of  no value in our analysis, whereas province/state is variable indicating a region within larger countries.Countries of interest are slovakia and Norway and we have no addotional data on specific locations within those countries.

In [55]:
df_cases.drop(['Province/State', 'Long','Lat',], axis = 1, inplace = True)

Checking if there are some missing values in the dataset:
df_cases_filtered.isna()

In [57]:

print(df_cases,df_cases.shape)
print('', df_cases.isna().any().sum())

  Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0         Norway        0        0        0        0        0        0   
1       Slovakia        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0   

I correctly dropped all the unneccessarry rows and dataset is ready to melt. But before I will do it I will prepare datasets with deaths and recovered cases.

## 4.2 Inspecting cleaning and preparing the dataset with deaths

### 4.2.1 Inspecting the dataset with deaths

In [61]:
print(df_deaths.shape)

(274, 449)


The deaths dataset has the same number of rows(274) and columns (449) as the cases dataset. its a good indication because the process of cleaning checking and extracting of data will be very simmilar.

In [63]:
print(df_deaths.head())

  Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  \
0       0      

Again, earlier dates indicate zero dates whereas number of deaths towards the later period is higher and steadily increasing Whereas first five variables: 'Province/State' 'Country/Region 'Lat''Long' seem to be identical through all the datasets. I have to confirm this assumption

I am checking what datatypes exist in the dataset deaths

In [66]:
df_deaths.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 274 entries, 0 to 273
Columns: 449 entries, Province/State to 4/10/21
dtypes: float64(2), int64(445), object(2)
memory usage: 961.3+ KB


df_deaths dataset seem to have identical types  two float columns (lat, long), two object columns (Province/State Country/Region) and 445 integer(date columns). I will verify it

In [68]:
print(df_deaths['Country/Region'].dtype)
print(df_deaths['Country/Region'].dtype)
print(df_deaths['Long'].dtype)
print(df_deaths['Lat'].dtype)

object
object
float64
float64


I confirm my assumption about float and object columns, I will check the date(integer columns as well)

In [70]:
#Filtering dataset so there are only dates
df_deaths_only_date = df_deaths.iloc[:, 4:]
#Creating loop to check which type the column is
date_counter = 0
for col in df_deaths_only_date:
    print(df_deaths_only_date[col].dtype)
    date_counter +=1 

print(date_counter)
print(df_deaths_only_date.info())

int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int64
int6

I confirm that the dataset contains 445 integers! I am finished with scrutinizing the dataset with deaths and I move further to clean it.

### 4.2.2 Cleaning the dataset with deaths

I first check whether columns contain missing values

In [74]:
print(df_deaths.isna().any())
df_deaths.isna().any().sum()



Province/State     True
Country/Region    False
Lat                True
Long               True
1/22/20           False
                  ...  
4/6/21            False
4/7/21            False
4/8/21            False
4/9/21            False
4/10/21           False
Length: 449, dtype: bool


3

#I will check how many columns with missing values are present. if there are three we know they are those:
 Province/State      
 Lat       
 Long

In [76]:
df_deaths.isna().any().sum()

3

I confirm the assumption about the missing values. Now I check identifier columns

In [78]:
print(df_deaths['Province/State'],df_deaths['Country/Region'])

0                                               NaN
1                                               NaN
2                                               NaN
3                                               NaN
4                                               NaN
5                                               NaN
6                                               NaN
7                                               NaN
8                      Australian Capital Territory
9                                   New South Wales
10                               Northern Territory
11                                       Queensland
12                                  South Australia
13                                         Tasmania
14                                         Victoria
15                                Western Australia
16                                              NaN
17                                              NaN
18                                              NaN
19          

I will drop the Long, Lat and Province/State columns as they are of no interest for us:

In [80]:
df_deaths.drop(['Province/State', 'Long', 'Lat'], axis = 1, inplace = True)

Filtering rows with Slovakia and Norway and re-writing original dataset:

In [82]:
df_deaths.head()

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0    Afghanistan        0        0        0        0        0        0   
1        Albania        0        0        0        0        0        0   
2        Algeria        0        0        0        0        0        0   
3        Andorra        0        0        0        0        0        0   
4         Angola        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   
2        0        0        0        0       0       0       0       0       0   
3        0        0        0        0       0       0       0       0       0   
4        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   
2       0       0       0       0        0        0        0        0   
3       0       0       0       0        0        0        0        0   
4       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0        0        0        0        0        0        0        0        0   
1        0        1        1        1        1        1        1        1   
2        0        0        1        2        3        4        4        4   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   3/18/20  3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  \
0        0        0        0        0        1        1        1        2   
1        2        2        2        2        2        4        5        5   
2        7        9       11       15       17       17       19       21   
3        0        0        0        0        1        1        1        1   
4        0        0        0        0        0        0        0        0   

   3/26/20  3/27/20  3/28/20  3/29/20  3/30/20  3/31/20  4/1/20  4/2/20  \
0        3        3        4        4        4        4       4       4   
1        6        8       10       10       11       15      15      16   
2       25       26       29       31       35       44      58      86   
3        3      

In [84]:
#row-filtering
df_deaths_filtered = """
SELECT * FROM df_deaths
WHERE "Country/Region" = 'Slovakia' OR "Country/Region" = 'Norway' """
#re-writing df_deaths
df_deaths = duckdb.query(df_deaths_filtered).to_df()

In [85]:

print(' The dataset has', df_deaths.isna().any().sum(), 'missings', df_deaths.shape, df_deaths.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Columns: 446 entries, Country/Region to 4/10/21
dtypes: int64(445), object(1)
memory usage: 7.1+ KB
 The dataset has 0 missings (2, 446) None


In [86]:
print(df_deaths)

  Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0       Slovakia        0        0        0        0        0        0   
1         Norway        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0   

 ##### Dataset is correctly cleaned and filtered with dates as integers, identifier column som object and zero missing values.

## 4.3 Inspecting cleaning and preparing the dataset with recovered cases

### 4.3.1 Inspecting the dataset with recovered cases

In [90]:
df_recovered.head()

Province/State Country/Region       Lat       Long  1/22/20  1/23/20  \
0            NaN    Afghanistan  33.93911  67.709953        0        0   
1            NaN        Albania  41.15330  20.168300        0        0   
2            NaN        Algeria  28.03390   1.659600        0        0   
3            NaN        Andorra  42.50630   1.521800        0        0   
4            NaN         Angola -11.20270  17.873900        0        0   

   1/24/20  1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   
2       0       0       0       0       0       0       0       0       0   
3       0       0       0       0       0       0       0       0       0   
4       0       0       0       0       0       0       0       0       0   

   2/10/20  2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/18/20  2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   
2        0        0        0        0        0        0        0        0   
3        0        0        0        0        0        0        0        0   
4        0        0        0        0        0        0        0        0   

   2/26/20  2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   
2        0        0        0        0       0       0       0       0       0   
3        0        0        0        0       0       0       0       0       0   
4        0        0        0        0       0       0       0       0       0   

   3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   
2       0       0       0       0        0        0        8        8   
3       0       0       0       0        0        0        1        0   
4       0       0       0       0        0        0        0        0   

   3/14/20  3/15/20  3/16/20  3/17/20  3/18/20  3/19/20  3/20/20  3/21/20  \
0        0        0        1        1        1        1        1        1   
1        0        0        0        0        0        0        0        2   
2       12       12       12       12       12       32       32       32   
3        1        1        1        1        1        1        1        1   
4        0        0        0        0        0        0        0        0   

   3/22/20  3/23/20  3/24/20  3/25/20  3/26/20  3/27/20  3/28/20  3/29/20  \
0        1        1        1        2        2        2        2        2   
1        2        2       10       17       17       31       31       33   
2       65       65       24       65       29       29       31       31   
3       

The dataset seems to follow the exactly same pattern as the two  datasets above.

In [92]:
df_recovered.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Columns: 449 entries, Province/State to 4/10/21
dtypes: float64(2), int64(445), object(2)
memory usage: 908.7+ KB


In [93]:
df_recovered.dtypes

Province/State     object
Country/Region     object
Lat               float64
Long              float64
1/22/20             int64
                   ...   
4/6/21              int64
4/7/21              int64
4/8/21              int64
4/9/21              int64
4/10/21             int64
Length: 449, dtype: object

Contrary to other columns variable lat is coded as an object. It doesn´t have an effect on my decision to drop it, it is just peculiar observation. Now I check whether there is  Norway and Slovakia in country/region variable.

In [95]:
#Checking country column
print(df_recovered['Country/Region'].isin(['Norway']).any())
print(df_recovered['Country/Region'].isin(['Slovakia']).any())


#Checking state column

print(df_recovered["Province/State"].isin(["Slovakia"]).any())
print(df_recovered['Province/State'].isin(['Norway']).any())

True
True
False
False


We did find both countries in Country/Region, Whereas we didnt find any of the countries of interest in Province/State. This means again that identifier will be Country/Region. I will confirm that dates are integers in subsequent code:

In [97]:
recovered_dates = df_recovered.iloc[:,4:]

In [98]:
recovered_dates.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 259 entries, 0 to 258
Columns: 445 entries, 1/22/20 to 4/10/21
dtypes: int64(445)
memory usage: 900.6 KB


All dates are integers

### 4.3.2 Cleaning the dataset with recovered cases

I will filter the rows dataset applying boolean mask to relevant rows and columns:

In [102]:
df_recoveredd =df_recovered[(df_recovered['Country/Region'] == 'Slovakia') | (df_recovered['Country/Region'] == 'Norway')].copy()

In [103]:
#Reassigning the new table to the old one
df_recovered = df_recoveredd
df_recovered

Province/State Country/Region     Lat     Long  1/22/20  1/23/20  1/24/20  \
186            NaN         Norway  60.472   8.4689        0        0        0   
212            NaN       Slovakia  48.669  19.6990        0        0        0   

     1/25/20  1/26/20  1/27/20  1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  \
186        0        0        0        0        0        0        0       0   
212        0        0        0        0        0        0        0       0   

     2/2/20  2/3/20  2/4/20  2/5/20  2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  \
186       0       0       0       0       0       0       0       0        0   
212       0       0       0       0       0       0       0       0        0   

     2/11/20  2/12/20  2/13/20  2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  \
186        0        0        0        0        0        0        0        0   
212        0        0        0        0        0        0        0        0   

     2/19/20  2/20/20  2/21/20  2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  \
186        0        0        0        0        0        0        0        0   
212        0        0        0        0        0        0        0        0   

     2/27/20  2/28/20  2/29/20  3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  \
186        0        0        0       0       0       0       0       0   
212        0        0        0       0       0       0       0       0   

     3/6/20  3/7/20  3/8/20  3/9/20  3/10/20  3/11/20  3/12/20  3/13/20  \
186       0       0       0       1        1        1        1        1   
212       0       0       0       0        0        0        0        0   

     3/14/20  3/15/20  3/16/20  3/17/20  3/18/20  3/19/20  3/20/20  3/21/20  \
186        1        1        1        1        1        1        1        1   
212        0        0        0        0        0        0        0        0   

     3/22/20  3/23/20  3/24/20  3/25/20  3/26/20  3/27/20  3/28/20  3/29/20  \
186        1        1        6        6        6        6        7        7   
212        7        7        7        7        2        2        2        2   

     3/30/20  3/31/20  4/1/20  4/2/20  4/3/20  4/4/20  4/5/20  4/6/20  4/7/20  \
186       12       13      13      32      32      32      32      32      32   
212        7        3       3       5      10      10      10       8      13   

     4/8/20  4/9/20  4/10/20  4/11/20  4/12/20  4/13/20  4/14/20  4/15/20  \
186      32      32       32       32       32       32       32       32   
212      16      23       23       23       23      107      113      151   

     4/16/20  4/17/20  4/18/20  4/19/20  4/20/20  4/21/20  4/22/20  4/23/20  \
186       32       32       32       32       32       32       32       32   
212      167      175      213      229      251      258      284      288   

     4/24/20  4/25/20  4/26/20  4/27/20  4/28/20  4/29/20  4/30/20  5/1/20  \
186       32       32       32       32       32       32       32      32   
212      355      386      394      403      423      484      524     558   

     5/2/20  5/3/20  5/4/20  5/5/20  5/6/20  5/7/20  5/8/20  5/9/20  5/10/20  \
186      32      32      32      32      32      32      32      32       32   
212     608     619     643     741     762     806     905     919      941   

     5/11/20  5/12/20  5/13/20  5/14/20  5/15/20  5/16/20  5/17/20  5/18/20  \
186       32       32       32       32       32       32       32       32   
212      959      983     1060     1112     1131     1151     1163     1185   

     5/19/20  5/20/20  5/21/20  5/22/20  5/23/20  5/24/20  5/25/20  5/26/20  \
186       32       32       32     7727     7727     7727     7727     7727   
212     1192     1231     1245     1256     1280     1301     1307     1322   

     5/27/20  5/28/20  5/29/20  5/30/20  5/31/20  6/1/20  6/2/20  6/3/20  \
186     7727     7727     7727     7727     7727    7727    7727    7727   
212     1327     1332     1338     1356     1366    1368    1372 

I need to change row indexes. I reset them.

In [105]:
df_recovered.reset_index(drop = True, inplace = True)

Now I  drop all the irelevant columns

In [107]:
#Droping unneccessarry columns
df_recovered.drop(['Long', 'Lat', 'Province/State'], axis = 1, inplace = True)
df_recovered

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0         Norway        0        0        0        0        0        0   
1       Slovakia        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0       0       0       0       0       0       0       0       0       1   
1       0       0       0       0       0       0       0       0       0   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0        1        1        1        1        1        1        1        1   
1        0        0        0        0        0        0        0        0   

   3/18/20  3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  \
0        1        1        1        1        1        1        6        6   
1        0        0        0        0        7        7        7        7   

   3/26/20  3/27/20  3/28/20  3/29/20  3/30/20  3/31/20  4/1/20  4/2/20  \
0        6        6        7        7       12       13      13      32   
1        2        2        2        2        7        3       3       5   

   4/3/20  4/4/20  4/5/20  4/6/20  4/7/20  4/8/20  4/9/20  4/10/20  4/11/20  \
0      32      32      32      32      32      32      32       32       32   
1      10      10      10       8      13      16      23       23       23   

   4/12/20  4/13/20  4/14/20  4/15/20  4/16/20  4/17/20  4/18/20  4/19/20  \
0       32       32       32       32       32       32       32       32   
1       23      107      113      151      167      175      213      229   

   4/20/20  4/21/20  4/22/20  4/23/20  4/24/20  4/25/20  4/26/20  4/27/20  \
0       32       32       32       32       32       32       32       32   
1      251      258      284      288      355      386      394      403   

   4/28/20  4/29/20  4/30/20  5/1/20  5/2/20  5/3/20  5/4/20  5/5/20  5/6/20  \
0       32       32       32      32      32      32      32      32      32   
1      423      484      524     558     608     619     643     741     762   

   5/7/20  5/8/20  5/9/20  5/10/20  5/11/20  5/12/20  5/13/20  5/14/20  \
0      32      32      32       32       32       32       32       32   
1     806     905     919      941      959      983     1060     1112   

   5/15/20  5/16/20  5/17/20  5/18/20  5/19/20  5/20/20  5/21/20  5/22/20  \
0       32       32       32       32       32       32       32     7727   
1     1131     1151     1163     1185     1192     1231     1245     1256   

   5/23/20  5/24/20  5/25/20  5/26/20  5/27/20  5/28/20  5/29/20  5/30/20  \
0     7727     7727     7727     7727     7727     7727     7727     7727   
1     1280     1301     1307     1322     1327     1332     1338     1356   

   5/31/20  6/1/20  6/2/20  6/3/20  6/4/20  6/5/20  6/6/20  6/7/20  6/8/20  \
0     7727    7727    7727    7727    8138    8138    8138    8138    8138   
1     1366    1368    1372    1375    1376    1379    1379    1389    1396   

   6/9/20  6/10/20  6/11/20  6/12/20  6/13/20  6/14/20  6/15/20 

##### -----Cleaning successfull! -----

## 4.4 Checking the tables all together

In [110]:
df_cases

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0         Norway        0        0        0        0        0        0   
1       Slovakia        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0        0        0        0        0        1        1        6       15   
1        0        0        0        0        0        0        0        0   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0      19      25      32      56      87     108     147     176     205   
1       0       0       0       0       0       1       1       3       3   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0      400      598      702      996     1090     1221     1333     1463   
1        7       10       16       32       44       54       63       72   

   3/18/20  3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  \
0     1550     1746     1914     2118     2385     2621     2863     3084   
1      105      123      137      178      185      186      204      216   

   3/26/20  3/27/20  3/28/20  3/29/20  3/30/20  3/31/20  4/1/20  4/2/20  \
0     3369     3755     4015     4284     4445     4641    4863    5147   
1      226      269      292      314      336      363     400     426   

   4/3/20  4/4/20  4/5/20  4/6/20  4/7/20  4/8/20  4/9/20  4/10/20  4/11/20  \
0    5370    5550    5687    5865    6086    6086    6211     6314     6409   
1     450     471     485     534     581     682     701      715      728   

   4/12/20  4/13/20  4/14/20  4/15/20  4/16/20  4/17/20  4/18/20  4/19/20  \
0     6525     6603     6623     6740     6896     6937     7036     7078   
1      742      769      835      863      977     1049     1089     1161   

   4/20/20  4/21/20  4/22/20  4/23/20  4/24/20  4/25/20  4/26/20  4/27/20  \
0     7156     7191     7338     7401     7463     7499     7527     7599   
1     1173     1199     1244     1325     1360     1373     1379     1381   

   4/28/20  4/29/20  4/30/20  5/1/20  5/2/20  5/3/20  5/4/20  5/5/20  5/6/20  \
0     7660     7710     7738    7783    7809    7847    7904    7955    7996   
1     1384     1391     1396    1403    1407    1408    1413    1421    1429   

   5/7/20  5/8/20  5/9/20  5/10/20  5/11/20  5/12/20  5/13/20  5/14/20  \
0    8034    8070    8099     8105     8132     8157     8175     8196   
1    1445    1455    1455     1457     1457     1465     1469     1477   

   5/15/20  5/16/20  5/17/20  5/18/20  5/19/20  5/20/20  5/21/20  5/22/20  \
0     8219     8237     8249     8257     8267     8281     8309     8332   
1     1480     1493     1494     1495     1495     1496     1502     1503   

   5/23/20  5/24/20  5/25/20  5/26/20  5/27/20  5/28/20  5/29/20  5/30/20  \
0     8346     8352     8364     8383     8401     8411     8422     8437   
1     1504     1509     1511     1513     1515     1520     1520     1521   

   5/31/20  6/1/20  6/2/20  6/3/20  6/4/20  6/5/20  6/6/20  6/7/20  6/8/20  \
0     8440    8446    8455    8477    8504    8522    8531    8547    8561   
1     1521    1522    1522    1525    1526    1526    1528    1528    1530   

   6/9/20  6/10/20  6/11/20  6/12/20  6/13/20  6/14/20  6/15/20 

In [111]:
df_deaths

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0       Slovakia        0        0        0        0        0        0   
1         Norway        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0       0       0       0       0       0       0       0       0       0   
1       0       0       0       0       0       0       0       0       0   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        3        3        3        3   

   3/18/20  3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  \
0        1        1        1        1        0        0        0        0   
1        6        7        7        7        7       10       12       14   

   3/26/20  3/27/20  3/28/20  3/29/20  3/30/20  3/31/20  4/1/20  4/2/20  \
0        0        0        0        0        0        0       1       1   
1       14       19       23       25       32       39      44      50   

   4/3/20  4/4/20  4/5/20  4/6/20  4/7/20  4/8/20  4/9/20  4/10/20  4/11/20  \
0       1       1       1       2       2       2       2        2        2   
1      59      62      71      76      89     101     108      113      119   

   4/12/20  4/13/20  4/14/20  4/15/20  4/16/20  4/17/20  4/18/20  4/19/20  \
0        2        2        2        6        8        9       11       12   
1      128      134      139      150      152      161      164      165   

   4/20/20  4/21/20  4/22/20  4/23/20  4/24/20  4/25/20  4/26/20  4/27/20  \
0       13       14       14       15       17       17       18       18   
1      181      182      187      194      199      201      201      205   

   4/28/20  4/29/20  4/30/20  5/1/20  5/2/20  5/3/20  5/4/20  5/5/20  5/6/20  \
0       20       22       23      23      24      24      25      25      25   
1      206      207      210     210     211     211     214     215     216   

   5/7/20  5/8/20  5/9/20  5/10/20  5/11/20  5/12/20  5/13/20  5/14/20  \
0      26      26      26       26       26       27       27       27   
1     217     218     219      219      224      228      229      232   

   5/15/20  5/16/20  5/17/20  5/18/20  5/19/20  5/20/20  5/21/20  5/22/20  \
0       27       28       28       28       28       28       28       28   
1      232      232      232      233      233      234      235      235   

   5/23/20  5/24/20  5/25/20  5/26/20  5/27/20  5/28/20  5/29/20  5/30/20  \
0       28       28       28       28       28       28       28       28   
1      235      235      235      235      235      236      236      236   

   5/31/20  6/1/20  6/2/20  6/3/20  6/4/20  6/5/20  6/6/20  6/7/20  6/8/20  \
0       28      28      28      28      28      28      28      28      28   
1      236     236     237     237     238     238     238     238     239   

   6/9/20  6/10/20  6/11/20  6/12/20  6/13/20  6/14/20  6/15/20 

In [112]:
df_recovered

Country/Region  1/22/20  1/23/20  1/24/20  1/25/20  1/26/20  1/27/20  \
0         Norway        0        0        0        0        0        0   
1       Slovakia        0        0        0        0        0        0   

   1/28/20  1/29/20  1/30/20  1/31/20  2/1/20  2/2/20  2/3/20  2/4/20  2/5/20  \
0        0        0        0        0       0       0       0       0       0   
1        0        0        0        0       0       0       0       0       0   

   2/6/20  2/7/20  2/8/20  2/9/20  2/10/20  2/11/20  2/12/20  2/13/20  \
0       0       0       0       0        0        0        0        0   
1       0       0       0       0        0        0        0        0   

   2/14/20  2/15/20  2/16/20  2/17/20  2/18/20  2/19/20  2/20/20  2/21/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   2/22/20  2/23/20  2/24/20  2/25/20  2/26/20  2/27/20  2/28/20  2/29/20  \
0        0        0        0        0        0        0        0        0   
1        0        0        0        0        0        0        0        0   

   3/1/20  3/2/20  3/3/20  3/4/20  3/5/20  3/6/20  3/7/20  3/8/20  3/9/20  \
0       0       0       0       0       0       0       0       0       1   
1       0       0       0       0       0       0       0       0       0   

   3/10/20  3/11/20  3/12/20  3/13/20  3/14/20  3/15/20  3/16/20  3/17/20  \
0        1        1        1        1        1        1        1        1   
1        0        0        0        0        0        0        0        0   

   3/18/20  3/19/20  3/20/20  3/21/20  3/22/20  3/23/20  3/24/20  3/25/20  \
0        1        1        1        1        1        1        6        6   
1        0        0        0        0        7        7        7        7   

   3/26/20  3/27/20  3/28/20  3/29/20  3/30/20  3/31/20  4/1/20  4/2/20  \
0        6        6        7        7       12       13      13      32   
1        2        2        2        2        7        3       3       5   

   4/3/20  4/4/20  4/5/20  4/6/20  4/7/20  4/8/20  4/9/20  4/10/20  4/11/20  \
0      32      32      32      32      32      32      32       32       32   
1      10      10      10       8      13      16      23       23       23   

   4/12/20  4/13/20  4/14/20  4/15/20  4/16/20  4/17/20  4/18/20  4/19/20  \
0       32       32       32       32       32       32       32       32   
1       23      107      113      151      167      175      213      229   

   4/20/20  4/21/20  4/22/20  4/23/20  4/24/20  4/25/20  4/26/20  4/27/20  \
0       32       32       32       32       32       32       32       32   
1      251      258      284      288      355      386      394      403   

   4/28/20  4/29/20  4/30/20  5/1/20  5/2/20  5/3/20  5/4/20  5/5/20  5/6/20  \
0       32       32       32      32      32      32      32      32      32   
1      423      484      524     558     608     619     643     741     762   

   5/7/20  5/8/20  5/9/20  5/10/20  5/11/20  5/12/20  5/13/20  5/14/20  \
0      32      32      32       32       32       32       32       32   
1     806     905     919      941      959      983     1060     1112   

   5/15/20  5/16/20  5/17/20  5/18/20  5/19/20  5/20/20  5/21/20  5/22/20  \
0       32       32       32       32       32       32       32     7727   
1     1131     1151     1163     1185     1192     1231     1245     1256   

   5/23/20  5/24/20  5/25/20  5/26/20  5/27/20  5/28/20  5/29/20  5/30/20  \
0     7727     7727     7727     7727     7727     7727     7727     7727   
1     1280     1301     1307     1322     1327     1332     1338     1356   

   5/31/20  6/1/20  6/2/20  6/3/20  6/4/20  6/5/20  6/6/20  6/7/20  6/8/20  \
0     7727    7727    7727    7727    8138    8138    8138    8138    8138   
1     1366    1368    1372    1375    1376    1379    1379    1389    1396   

   6/9/20  6/10/20  6/11/20  6/12/20  6/13/20  6/14/20  6/15/20 

All the tables seem to be correctly filtered, I confirm that by printing out datatypes in columns of respective dataframes.

In [114]:
print(df_recovered.info())
print(df_deaths.info())
print(df_cases.info())



<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Columns: 446 entries, Country/Region to 4/10/21
dtypes: int64(445), object(1)
memory usage: 7.1+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Columns: 446 entries, Country/Region to 4/10/21
dtypes: int64(445), object(1)
memory usage: 7.1+ KB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2 entries, 0 to 1
Columns: 446 entries, Country/Region to 4/10/21
dtypes: int64(445), object(1)
memory usage: 7.1+ KB
None


All the dataframes have 445 integers(dates) and 1 object(identifier). I will finally confirm a cleanliness of dataset viewing their shape

In [116]:
print(df_recovered.shape)
print(df_deaths.shape)
print(df_cases.shape)

(2, 446)
(2, 446)
(2, 446)


All of them have two rows(Slovakia and Norway) and 446 columns out of which 445 are date columns and 1 is identifier. I confirm dataset is ready for further analysis.

# 5 Melting the datasets

## 5.1 Melting the cases dataset

Melting means that I change format of the table. Instead of having 446 columns, I want to have just three. these are:
country/region, date and number of cases.The table will be long because each date column will be listed as row twice- one for slovakia and same date for norway. 

In [121]:
df_cases = df_cases.melt(id_vars = 'Country/Region', var_name = 'date', value_name = "cases_number").copy()

Checking the table

In [123]:
df_cases

Country/Region     date  cases_number
0           Norway  1/22/20             0
1         Slovakia  1/22/20             0
2           Norway  1/23/20             0
3         Slovakia  1/23/20             0
4           Norway  1/24/20             0
..             ...      ...           ...
885       Slovakia   4/8/21        368470
886         Norway   4/9/21        102719
887       Slovakia   4/9/21        369393
888         Norway  4/10/21        102974
889       Slovakia  4/10/21        370473

[890 rows x 3 columns]

 ##### The table is melted correctly, there are 890 rows which is 455 per country this number of rows corresponds to number of date columns in the original dataset

## 5.2 Melting the deaths dataset

In [126]:
df_deaths = df_deaths.melt(id_vars = "Country/Region", var_name = 'date', value_name = 'deaths_number').copy()

Checking the table

In [128]:
df_deaths

Country/Region     date  deaths_number
0         Slovakia  1/22/20              0
1           Norway  1/22/20              0
2         Slovakia  1/23/20              0
3           Norway  1/23/20              0
4         Slovakia  1/24/20              0
..             ...      ...            ...
885         Norway   4/8/21            684
886       Slovakia   4/9/21          10411
887         Norway   4/9/21            684
888       Slovakia  4/10/21          10487
889         Norway  4/10/21            684

[890 rows x 3 columns]

 ##### I confirm that columns are named correctly and shape of the table 890 x 3 is desirable.

## 5.3 Melting the dataset with recovered cases

In [131]:
df_recovered = df_recovered.melt(id_vars = "Country/Region", var_name = "date", value_name = 'recovered_number').copy()

Checking the dataframe

In [133]:
df_recovered

Country/Region     date  recovered_number
0           Norway  1/22/20                 0
1         Slovakia  1/22/20                 0
2           Norway  1/23/20                 0
3         Slovakia  1/23/20                 0
4           Norway  1/24/20                 0
..             ...      ...               ...
885       Slovakia   4/8/21            255300
886         Norway   4/9/21             17998
887       Slovakia   4/9/21            255300
888         Norway  4/10/21             17998
889       Slovakia  4/10/21            255300

[890 rows x 3 columns]

I confirm that table shape matches perfectly the table shapes of recovered and cases tables. All three tables begin and finish with the same date. Lastly i check if there are any duplicates in the date columns. Each dataset should not have more than 2 duplicates of date(corresponding to 2 respective countries) in itself.

Accessing the number of duplicates for each date in each dataset(correct should be 2 for each date)

In [136]:
cases_duplicates_size = df_cases.groupby("date").size()
deaths_duplicates_size = df_deaths.groupby("date").size()
recovered_duplicates_size = df_recovered.groupby("date").size()

Checking if each date contains no more or no less than two duplicates:

In [138]:

print((cases_duplicates_size == 2).all())
print((deaths_duplicates_size == 2).all())
print((recovered_duplicates_size == 2).all())

True
True
True


 ##### Melting successfull!

# 6. Merging datasets

I will merge the datasets using "LEFT JOINs". Since the individual datasets ("df_cases", "df_deaths", "df_recovered") are already clean and perfectly aligned with 890 rows each for Norway and Slovakia over time, the final merged dataset should also retain this length (890 rows) if the joins are performed correctly.

While two simple "LEFT JOIN"s would suffice for this specific merging task, I want to demonstrate proficiency in more in-demand SQL skills for my portfolio. Therefore, I will structure these joins using a Common Table Expression (CTE) and a subquery. This approach showcases my ability to write modular, readable, and more complex SQL queries, which is a valuable skill for data analysis roles.

In [142]:
merging_query = """ WITH recovered_join  AS (SELECT 
                            "Country/Region",
                            "date", 
                            "recovered_number" 
FROM df_recovered)

SELECT c."Country/Region" AS country,
       c."date", 
       c."cases_number", 
       d."deaths_number", 
       r."recovered_number"
FROM df_cases AS c

LEFT JOIN (SELECT "deaths_number", 
                  "date", 
                  "Country/Region"
           FROM df_deaths) AS d
on c."date" = d."date" AND c."Country/Region" = d."Country/Region"
              
LEFT JOIN recovered_join AS r 
on c."date" = r."date" AND r."Country/Region" = c."Country/Region";  """

Turning query into dataframe

In [144]:
df_timeseries = duckdb.query(merging_query).to_df()
df_timeseries 

country      date  cases_number  deaths_number  recovered_number
0      Norway   1/23/20             0              0                 0
1    Slovakia   1/23/20             0              0                 0
2      Norway   1/24/20             0              0                 0
3    Slovakia   1/24/20             0              0                 0
4      Norway   1/25/20             0              0                 0
..        ...       ...           ...            ...               ...
885  Slovakia  10/22/20         35330            115              8763
886  Slovakia   4/20/20          1173             13               251
887    Norway   8/10/20          9684            256              8857
888  Slovakia   8/10/20          2599             31              1866
889    Norway  10/22/20         17234            279             11863

[890 rows x 5 columns]

## 6.1 Verifying integrity of the time series dataframe

Table df_timeseries  contains 890 rows and five columns which is expected outcome of merging.Finally I check missing values. if there are none it means merging was succesfull

In [147]:
print(df_timeseries .isnull().any())
print(df_timeseries.info())

country             False
date                False
cases_number        False
deaths_number       False
recovered_number    False
dtype: bool
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 890 entries, 0 to 889
Data columns (total 5 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   country           890 non-null    object
 1   date              890 non-null    object
 2   cases_number      890 non-null    int64 
 3   deaths_number     890 non-null    int64 
 4   recovered_number  890 non-null    int64 
dtypes: int64(3), object(2)
memory usage: 34.9+ KB
None


##### ----No missing values found so merging went well----
I see that date is coded as object and I need to convert it to date-time format

# 6. Date-time conversion

I pluck first row with date column first

In [151]:
df_timeseries

country      date  cases_number  deaths_number  recovered_number
0      Norway   1/23/20             0              0                 0
1    Slovakia   1/23/20             0              0                 0
2      Norway   1/24/20             0              0                 0
3    Slovakia   1/24/20             0              0                 0
4      Norway   1/25/20             0              0                 0
..        ...       ...           ...            ...               ...
885  Slovakia  10/22/20         35330            115              8763
886  Slovakia   4/20/20          1173             13               251
887    Norway   8/10/20          9684            256              8857
888  Slovakia   8/10/20          2599             31              1866
889    Norway  10/22/20         17234            279             11863

[890 rows x 5 columns]

In [152]:
print(type((df_timeseries.iloc[0, 1])))
print(df_timeseries.loc[0, "date"])

<class 'str'>
1/23/20


It is string format, but i need date - time format. I continue with  changing  cariable string format to date-time

In [154]:
df_timeseries["date"] = pd.to_datetime(df_timeseries["date"], format = "%m/%d/%y")

In [155]:
df_timeseries

country       date  cases_number  deaths_number  recovered_number
0      Norway 2020-01-23             0              0                 0
1    Slovakia 2020-01-23             0              0                 0
2      Norway 2020-01-24             0              0                 0
3    Slovakia 2020-01-24             0              0                 0
4      Norway 2020-01-25             0              0                 0
..        ...        ...           ...            ...               ...
885  Slovakia 2020-10-22         35330            115              8763
886  Slovakia 2020-04-20          1173             13               251
887    Norway 2020-08-10          9684            256              8857
888  Slovakia 2020-08-10          2599             31              1866
889    Norway 2020-10-22         17234            279             11863

[890 rows x 5 columns]

looks correct but I will sort it by date and check type of date column once again.

In [157]:
print(type(df_timeseries.loc[0, "date"]))

<class 'pandas._libs.tslibs.timestamps.Timestamp'>


It is date-time format! Now i sort the  dataset based on date

In [159]:
df_timeseries.sort_values("date")

country       date  cases_number  deaths_number  recovered_number
741    Norway 2020-01-22             0              0                 0
742  Slovakia 2020-01-22             0              0                 0
0      Norway 2020-01-23             0              0                 0
1    Slovakia 2020-01-23             0              0                 0
2      Norway 2020-01-24             0              0                 0
..        ...        ...           ...            ...               ...
738  Slovakia 2021-04-08        368470          10322            255300
739    Norway 2021-04-09        102719            684             17998
740  Slovakia 2021-04-09        369393          10411            255300
864  Slovakia 2021-04-10        370473          10487            255300
841    Norway 2021-04-10        102974            684             17998

[890 rows x 5 columns]

##### ---Dataset is sorted correctly and ready for explanatory data analysis(EDA)---

#7 Explanatory data analysis

Rolling AVGS etc.